📈 Capture onboarding funnel

shivekkhurana opened the pull request:
This PR updates Andrea's work to capture all views of the onboarding flow. 

It also updates the method to capture events. Instead of specifying verbs for each event, we capture all view events by screen id. 

Non view events still need a verb. 

Also fixes: #20664

shivekkhurana committed:
📈 Capture onboarding funnel

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (20)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 136f6906 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/1/) | 2024-07-30 17:03:22 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/1/consoleText) |
| :heavy_check_mark: | 136f6906 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-20926/1/) | 2024-07-30 17:07:10 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-170039-136f69-pr20926-x86_64.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-170039-136f69-pr20926-x86_64.apk)|
| :heavy_check_mark: | 136f6906 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-20926/1/) | 2024-07-30 17:08:28 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-170039-136f69-pr20926-arm64-v8a.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-170039-136f69-pr20926-arm64-v8a.apk)|
| :heavy_check_mark: | 136f6906 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-20926/1/) | 2024-07-30 17:12:33 | ~11 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/mnQgMq) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2FmnQgMq)|
| | | | | | | |
| :x: | 46fda9bc | [#2](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/2/) | 2024-07-30 18:26:22 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/2/consoleText) |
| :heavy_check_mark: | 46fda9bc | [#2](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-20926/2/) | 2024-07-30 18:31:15 | ~7 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-182354-46fda9-pr20926-x86_64.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-182354-46fda9-pr20926-x86_64.apk)|
| :heavy_check_mark: | 46fda9bc | [#2](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-20926/2/) | 2024-07-30 18:31:33 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-182354-46fda9-pr20926-arm64-v8a.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-182354-46fda9-pr20926-arm64-v8a.apk)|
| :heavy_check_mark: | 46fda9bc | [#2](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-20926/2/) | 2024-07-30 18:36:41 | ~12 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/jrSS64) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2FjrSS64)|
| | | | | | | |
| :x: | 29dbd178 | [#3](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/3/) | 2024-07-30 18:56:15 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/3/consoleText) |
| :heavy_check_mark: | 29dbd178 | [#3](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-20926/3/) | 2024-07-30 19:00:46 | ~7 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-185330-29dbd1-pr20926-x86_64.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-185330-29dbd1-pr20926-x86_64.apk)|
| :heavy_check_mark: | 29dbd178 | [#3](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-20926/3/) | 2024-07-30 19:00:52 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-185330-29dbd1-pr20926-arm64-v8a.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-185330-29dbd1-pr20926-arm64-v8a.apk)|
| :heavy_check_mark: | 29dbd178 | [#3](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-20926/3/) | 2024-07-30 19:09:51 | ~16 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/m3yPDE) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2Fm3yPDE)|
| | | | | | | |
| :heavy_check_mark: | 8db5a076 | [#4](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/4/) | 2024-07-30 20:23:15 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-201820-8db5a0-pr20926-tests.log) |
| :heavy_check_mark: | 8db5a076 | [#4](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-20926/4/) | 2024-07-30 20:26:52 | ~8 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-201815-8db5a0-pr20926-x86_64.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-201815-8db5a0-pr20926-x86_64.apk)|
| :heavy_check_mark: | 8db5a076 | [#4](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-20926/4/) | 2024-07-30 20:27:18 | ~8 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240730-201820-8db5a0-pr20926-arm64-v8a.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240730-201820-8db5a0-pr20926-arm64-v8a.apk)|
| :heavy_check_mark: | 8db5a076 | [#4](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-20926/4/) | 2024-07-30 20:30:29 | ~12 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/WLE3AA) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2FWLE3AA)|
| | | | | | | |
| :heavy_check_mark: | 52fef78a | [#5](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/5/) | 2024-07-31 07:51:09 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240731-074626-52fef7-pr20926-tests.log) |
| :heavy_check_mark: | 52fef78a | [#5](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-20926/5/) | 2024-07-31 07:52:49 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240731-074622-52fef7-pr20926-x86_64.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240731-074622-52fef7-pr20926-x86_64.apk)|
| :heavy_check_mark: | 52fef78a | [#5](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-20926/5/) | 2024-07-31 07:54:24 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240731-074626-52fef7-pr20926-arm64-v8a.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240731-074626-52fef7-pr20926-arm64-v8a.apk)|
| :heavy_check_mark: | 52fef78a | [#5](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-20926/5/) | 2024-07-31 07:59:04 | ~12 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/etgvNz) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2FetgvNz)|
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 56943548 | [#6](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/6/) | 2024-08-02 16:05:34 | ~4 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240802-160114-569435-pr20926-tests.log) |
| | | | | | | |
| :heavy_check_mark: | f70fa9f8 | [#7](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20926/7/) | 2024-08-02 16:10:18 | ~3 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240802-160618-f70fa9-pr20926-tests.log) |
| :heavy_check_mark: | f70fa9f8 | [#7](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-20926/7/) | 2024-08-02 16:13:14 | ~6 min | `android-e2e` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240802-160614-f70fa9-pr20926-x86_64.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240802-160614-f70fa9-pr20926-x86_64.apk)|
| :heavy_check_mark: | f70fa9f8 | [#7](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-20926/7/) | 2024-08-02 16:13:54 | ~7 min | `android` | [:robot:`apk`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240802-160618-f70fa9-pr20926-arm64-v8a.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-mobile-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-240802-160618-f70fa9-pr20926-arm64-v8a.apk)|
| :heavy_check_mark: | f70fa9f8 | [#7](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-20926/7/) | 2024-08-02 16:20:47 | ~14 min | `ios` | [:iphone:`ipa`](https://i.diawi.com/oce7gP) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2Foce7gP)|

ilmotta commented on src/status_im/contexts/centralized_metrics/tracking.cljs:
```diff
@@ -12,25 +12,38 @@
-    :chats-stack       (user-journey-event chats-tab-clicked)
-    :wallet-stack      (user-journey-event wallet-tab-clicked)
-    nil))
+  (when (some #{view-id} view-ids-to-track)
```
Probably not a problem @shivekkhurana, but `case` does a constant time dispatch, so it's much faster than `some` used with sets. Ideally in the future we could have a macro to make this more convenient but have the least amount of overhead possible.

But totally looks good for now!

shivekkhurana replied:
My idea was to have a list of screen ids that we can track.

But I understand that `some` is not the best. So I have used `contains?` with a Set which will have an O(1) complexity in the average case.

ilmotta reviewed (approved):
🚀

shivekkhurana committed:
⏩ Faster lookup

shivekkhurana committed:
✏️ Fix tests

seanstrom commented on src/status_im/contexts/centralized_metrics/tracking.cljs:
```diff
@@ -3,37 +3,61 @@
-  [action]
+(defn key-value-event
+  [event-name val-key value]
   {:metric
-   {:eventName  "user-journey"
+   {:eventName  event-name
     :platform   platform/os
     :appVersion build/app-short-version
-    :eventValue {:action action}}})
+    :eventValue {val-key value}}})
+
+(defn user-journey-event
+  [action]
+  (key-value-event "user-journey" :action action))
+
+(defn navigation-event
+  [view-id]
+  (key-value-event "navigation" :viewId view-id))
```
Little code comment 🐭 
Could we pass the `val-key` and `value` as a map?
For example, we could call `key-value-event` like:
```clojure
(key-value-event "event-name" {:event-key event-value})
```
Perhaps if we do that we would rename `key-value-event` into `named-event` or maybe `metric-event`. Thoughts?

shivekkhurana replied:
It's better for sure. But I built it in a rush yesterday evening. 

We wanted to postpone it to 2.31, but Icaro suggested that getting onboarding in is not hard. 

This is only the first version. 2.31 will be way more elaborate.

seanstrom replied:
Okay sounds good, thanks for clarifying 🙏

seanstrom commented on src/status_im/contexts/centralized_metrics/tracking.cljs:
```diff
@@ -3,37 +3,61 @@
 (defn tracked-event
   [[event-name second-parameter]]
   (case event-name
-    :onboarding/navigate-to-create-profile
-    (user-journey-event navigate-to-create-profile-event)
-
```
Why are we removing this track event?

shivekkhurana replied:
Because it's being tracked a navigation event, so we don't need to track the click.

seanstrom reviewed (approved):
Thanks for the PR @shivekkhurana 
I left a few small comments, looks good ✅

churik commented:
@shivekkhurana 
can I smehow verify that it is captured?

shivekkhurana commented:
> @shivekkhurana can I smehow verify that it is captured?

Yes, goto mixpanel, login with your status email and look for events. Easiest ones to see will be navigating between wallet and communities tab.

churik commented:
yes, I found it, it was the first time for me checking it.

So for 1 onboarding flow, we capture at least 8 events:
![image](https://github.com/user-attachments/assets/ba8f6492-5103-40a5-8d96-69f2067f049e)

Are we sure that DB is ready for this amount of data? 

And as far as I understood every possible account creation path should be covered, right? 
I'll recheck this and recheck the app after upgrade.

churik commented:
Also all re-logins are captured (at least 3 events each) -do we need this data @shivekkhurana @ilmotta ?
(user journey > navigation > navigation, check the [link](https://eu.mixpanel.com/project/3350627/view/3857659/app/profile#distinct_id=9e7a8e40-2ed6-4171-b312-aea2f6098fb1) )

ilmotta commented:
> Also all re-logins are captured (at least 3 events each) -do we need this data

We don't need necessarily, but at least while 2.30 is our active release in stores, having more can be better than less. Capturing re-logins can give us some clues about where people are failing to get into the app.

Right now there's no concept of a session. Getting insights from 2.30 data will be hard.

> Are we sure that DB is ready for this amount of data?

We are vulnerable to spam, like in many other parts of the product. If an actor wants, they could cause huge spikes and distort metrics. The current data plan should be sufficient for 2.30 (@cammellos knows better this part)

churik commented:
Thank you. In this case, this PR is ready to be merged and cherry-picked. Thank you for explanation and instructions @shivekkhurana @ilmotta 
Tests can be skipped, as it doesn't affect them.

shivekkhurana committed:
✅ Capture how many people enabled metrics

- It technically captures disabled too
- But we'll never know if someone disabled
- Because that info won't be transmitted

shivekkhurana committed:
🧯 Fix lint

shivekkhurana committed:
Merge branch 'develop' into shivekkhurana/onboarding-metrics
